doc/stdenv.xml

```diff
@@ -179,68 +179,252 @@ genericBuild
+  <itemizedlist>
+    <listitem>
+      <para>
+      The type of dependency is determined by the sum of the offsets assigned to each edge of the path
+      </para>
+    </listitem>
+    <listitem>
+      <para>
+      The path is only included if the subpaths sharing the starting point never go out of bounds (<literal>&lt; -2</literal>, <literal>&gt; 1</literal>).
+      This is a fancy way of saying we prune as we go.
+      </para>
+    </listitem>
+  </itemizedlist>
+  Formally:
+  <programlisting>
+propagated-dep(h0, _, A, B)
+propagated-dep(h1, t1, B, C)
+h0 + h1 in {-1, 0, 1}
+h0 + t1 in {-1, 0, 1}
+-------------------------------------- Transitive property
+propagated-dep(h0 + h1, h0 + t1, A, C)</programlisting>
+  <programlisting>
+dep(h0, _, A, B)
+propagated-dep(h1, t1, B, C)
+h0 + h1 in {-2, -1, 0, 1}
+h0 + t1 in {-1, 0, 1}
+-------------------------------------- Take immediate deps' propagated deps
+propagated-dep(h0 + h1, h0 + t1, A, C)</programlisting>
+  <programlisting>
+propagated-dep(h, t, A, B)
+-------------------------------------- Propagated deps count as deps
+dep(h, t, A, B)</programlisting>
+</para>
 
 <variablelist>
-  <title>Variables affecting <literal>stdenv</literal>
-  initialisation</title>
+  <title>Variables specifying dependencies</title>
 
-  <varlistentry>
-    <term><varname>NIX_DEBUG</varname></term>
-    <listitem><para>If set, <literal>stdenv</literal> will print some
-    debug information during the build.  In particular, the
-    <command>gcc</command> and <command>ld</command> wrapper scripts
-    will print out the complete command line passed to the wrapped
-    tools.</para></listitem>
+   <varlistentry>
+     <term><varname>depsBuildBuild</varname></term>
+     <listitem>
+       <para>
+         A list of dependencies whose host and target platforms are the new derivation's build platform.
+         This means a <literal>-1</literal> host and <literal>-1</literal> target offset from the new derivation's platforms.
+         They are programs/libraries used at build time that furthermore produce programs/libraries also used at build time.
+         If the dependency doesn't care about the target platform (i.e. isn't a compiler or similar tool), put it in <varname>nativeBuildInputs</varname>instead.
+         The most common use for this <literal>buildPackages.stdenv.cc</literal>, the default C compiler for this role.
+         That example crops up more than one might think in old commonly used C libraries.
+       </para>
+       <para>
+         Since these packages are able to be run at build time, that are always added to the <envar>PATH</envar>, as described above.
+         But since these packages are only guaranteed to be able to run then, they shouldn't persist as run-time dependencies.
+         This isn't currently enforced, but could be in the future.
+       </para>
+     </listitem>
   </varlistentry>
 
-</variablelist>
+  <varlistentry>
```

Comment:
Honest question: How are mere mortals supposed to make any sense of this? Why is all this needed?

In e.g. Buildroot, all that's needed is:
````
KMOD_DEPENDENCIES += zlib
````
and:
````
SDL_DEPENDENCIES += host-automake host-autoconf host-libtool
````

Nothing else. In the current cross build system the equivalents are `buildInputs` and `nativeBuildInputs`. Somehow now we need an absurd amount of new ones?

Reply:
I wish we didn't, but for compilers that build their own standard library we do. If you look at gcc, I use all three kinds of build dependencies. For GHC I expect to additionally use the `depsTargetTarget` and, eventually if cross compilers support compilers plugins, `depsHostHost` too---all 6 in total. For non-Canadian crosses, yes some of these collapse, but they collapse in different ways depending on which platforms coincide. Were we to support Canadian crosses, they would all be simultaneously unique.

In a glorious future where every compiler is multi-target and standard libraries and runtimes are built just like normal dependencies, I'll be happy to rip all this away. Also, for the vast majority of packages, `buildInputs` and `nativeBuildInputs` are all that's needed---just like today.

I plan on adding a warning to the documentation saying all the new dependencies are unstable, in addition to the `__`-prefixes I already added.